Rework of the Python Filter (Depends on PR Support UDP TCP #224)

ThibaudCartegnie opened the pull request:
# :sparkles: Pull Request Template
:bangbang: Once all the **checklist** is **done** you have to:
  * **stash merge** this pull request
  * **delete** the corresponding **branch**
  * **close** the associated **issue**

## :page_with_curl: Type of change

Please delete options that are not relevant.

**Breaking change**: fix or feature that would cause existing functionality to not work as expected.

## :bulb: Related Issue(s)

- Depends on #224 

## :black_nib: Description

Reworks the python filter to be closer of what we need and easier to use.

## :dart: Test Environments

### FreeBSD - Vulture ISO (12.2)
- Redis (6.2.6)
- Boost (1.72.0)
- clang++(10.0.1)
- CMake (3.21.3)
- Python (3.8.11)

### Ubuntu (20.04)
- Redis (5.0.7)
- Boost (1.71.0)
- g++ (9.3.0)
- CMake (3.16.3)
- Python (3.8.10)
- Valgrind (3.15.0)


## :heavy_check_mark: Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] (**If new filter**) I have added corresponding page to the documentation
- [ ] (**If other changes**) I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

</br>

- [ ] :raising_hand: **I certify on my honor that all the information provided is true, and I've done all I can to deliver a high-quality code**

Commit:
separated unix socket logic from generic logic in the server class into 2 classes

ThibaudCartegnie committed:
Separated logic between tasks and Sessions, the separation is only applied on the dga filter for now

ThibaudCartegnie committed:
Fixed potential corruption :
If more than 1 certitude is sent back to client AND a body is set, the body is copied on the memory location of the certitudes
It is fixed by using the correct offset when copying the body data.
The other function using similar logic has been modified to keep it coherent.

ThibaudCartegnie committed:
Separated Sessions into UnixSession and TcpSession
UnixSessions are spawn by UnixServer and TcpSessions by TcpServer
Included a modification when passing arguments from sessions to task (move instead of ref)

ThibaudCartegnie committed:
WIP : First draft of DarwinPacket struct
 - Used to pass needed data from Session to Task

ThibaudCartegnie committed:
Finished integration of DarwinPacket in session
(some work is still need for the tasks)

ThibaudCartegnie committed:
Modified DarwinPacket to be more OOP-like
State accessible by const getters only
State mutable with only two methods : add Certitude and mutateBody

ThibaudCartegnie committed:
Re-wired some things in the tasks
Added a few methods for DarwinPacket

ThibaudCartegnie committed:
Put back structs in c header
Added ParseBody in ATask::run with error handling

ThibaudCartegnie committed:
Adapted all compiled filters
Fixed issues with threadpool instance and size issues in the ASession

ThibaudCartegnie committed:
Added parsing of tcp/udp
Fixed issues in the serialization
Added support for ipv6 in Tcp sessions

ThibaudCartegnie committed:
Adapted TcpServer and Config for ip addresses
Adapted python manager to handle tcp sockets

ThibaudCartegnie committed:
DarwinPacket: fixed issues with the serialization
The packet was illformed near the certitudes
NextFilterConnector: Working WIP

ThibaudCartegnie committed:
Separated NextFilterConnector in :
Abstract/Tcp/Unix, fixed configuration for nextFilter

ThibaudCartegnie committed:
Moved all network related objects to network folder

ThibaudCartegnie committed:
Fixed Set of asio buffer in ANextFilterConnector

ThibaudCartegnie committed:
Fixed memory issues:
- DarwinPacket copied only the first 15 bytes of event id
- Destructor of AServer wasn't virtual leading to memory leak when
UnixServer was destroyed

ThibaudCartegnie committed:
Fixed problem with the manager
Service.update did not update nettwork field

ThibaudCartegnie committed:
Fixed manager (failed previous commit)

ThibaudCartegnie committed:
RedisManager: modified Rate Limit connection logic
Added retry attempts, modified tests accordingly

ThibaudCartegnie committed:
Adapted ATask, Asession and DarwinPacket
Proprietary filters needed additional methods for accessing private
fields of a packet

ThibaudCartegnie committed:
Adapted tests to TCP :
Added possibility to run all tests in TCP mode

ThibaudCartegnie committed:
Added tests for TCP connections
Assured working with ipv4 and ipv6 addresses
Fixed a few log lines in the redis test file

ThibaudCartegnie committed:
Added UDP protocol to darwin
Fixed a few problems and cleaned some code
Added tests for tcp and udp
Added possibility to run all test in unix or tcp

ThibaudCartegnie committed:
Merge branch 'dev' into support_tcp_udp

ThibaudCartegnie committed:
fixed few issues with the manager

ThibaudCartegnie committed:
Added comments documentation
Removed unused methods

ThibaudCartegnie committed:
Fixed method names

ThibaudCartegnie committed:
Few modification of NextFilter handling
Specification of move semantic for DarwinPacket

ThibaudCartegnie committed:
Few fixes when passing information to next filters

ThibaudCartegnie committed:
Added a tests for multi filters using the manager
Fixed a fex things accordingly

ThibaudCartegnie committed:
Fixed issue with UUIDs
EvtIdToString passed from Session to DarwinPacket

ThibaudCartegnie committed:
Fixed problems with next filters, cleaned tests
Also cleaned ASession a bit

ThibaudCartegnie committed:
Modified config of next filter for easier configuration

ThibaudCartegnie committed:
Modified protocol.h
added packed attribute to darwin header
Removed certitude list default size
Modified parsing and serializing accordingly

ThibaudCartegnie committed:
Fixed clang/gcc warnings
Most case were unused variables or poorly chosen types

ThibaudCartegnie committed:
Fixed compilation issues on hardenedBSD
Fixed issue with positional flags between test/manager and filters

ThibaudCartegnie committed:
Fixed last issue with UDP NextFilterConnector

ThibaudCartegnie committed:
Added Thread Pool Cpp dependency

ThibaudCartegnie committed:
First working version

ThibaudCartegnie committed:
Modified function structures, added missing links

ThibaudCartegnie committed:
WIP : fixed few things in the protocol python - Shared Lib

ThibaudCartegnie committed:
WIP : fixed return of operator (not completely)

ThibaudCartegnie committed:
Merge branch 'support_tcp_udp' into python_filter

ThibaudCartegnie committed:
Adapted filter for support_udp_tcp

ThibaudCartegnie committed:
WIP working threads, memory issues

ThibaudCartegnie committed:
Fixed issues with memory usage, added config step

ThibaudCartegnie committed:
Fixed memory problems with python, meory usage is now stable

ThibaudCartegnie committed:
Added configuration and contextualization phases

ThibaudCartegnie committed:
Refactored FunctionPySo and functions loading methods

ThibaudCartegnie committed:
Added possibility to log through darwin

frikilax commented on tests/filters/fpython.py:
```diff
@@ -0,0 +1,40 @@
+                        '"python_script_path": "{python_path}",\n' \
+                        '"shared_library_path": "{so_path}",\n' \
+                        '"log_file_path": "/tmp/python_alerts.log"\n' \
+                    '}}'.format(python_path='/home/myadvens.lan/tcartegnie/workspace/darwin/samples/fpython/example.py',
+                                so_path='/home/myadvens.lan/tcartegnie/workspace/pycpp/build/libfpython.so')
```
You need to variabilize these paths

ThibaudCartegnie committed:
Fix Compile issues : added cmake findPython module
example.py is made a better example to test the filter
Added logger to example
added support for venv
updated cmake min version

ThibaudCartegnie committed:
Moved example.py

ThibaudCartegnie committed:
Added tests and corrections

ThibaudCartegnie committed:
Changed structure of expected python script
- now expects a class (see algorithm_module.py)
- added tests for class specific issues

ThibaudCartegnie committed:
Resolve comments

ThibaudCartegnie commented on samples/fpython/Generator.cpp:
```diff
@@ -0,0 +1,457 @@
+    oss << "Python error '" << typeString << "' : " << errString;
+    log.log(log_type, oss.str());
+    return true;
+}
```
I added a check just after the PyErr_Fetch

ThibaudCartegnie commented on samples/fpython/PythonTask.cpp:
```diff
@@ -0,0 +1,358 @@
+            }
+
+            if ((_parsed_body = PyObject_CallFunctionObjArgs(_functions.parseBodyFunc.py, *raw_body, nullptr)) == nullptr) {
+                DARWIN_LOG_DEBUG("PythonTask:: ParseBody:: An error occurred while calling the Python function");
```
Set to ERROR

ThibaudCartegnie commented on samples/fpython/PythonTask.cpp:
```diff
@@ -0,0 +1,358 @@
+        case FunctionOrigin::shared_library:{
+            try{
+                if((_parsed_body = _functions.parseBodyFunc.so(_pModule, _packet.GetBody()))== nullptr){
+                    DARWIN_LOG_DEBUG("PythonTask:: ParseBody:: An error occurred while calling the SO function");
```
Set to ERROR

ThibaudCartegnie commented on samples/fpython/PythonTask.cpp:
```diff
@@ -0,0 +1,358 @@
+        }
+    }
+    for(auto cert: resp.certitudes) {
+        _packet.AddCertitude(cert);
```
I added it
